ceed-basis.c

```diff
@@ -290,46 +290,138 @@ int CeedQRFactorization(CeedScalar *mat, CeedInt m, CeedInt n) {
 
   for (i=0; i<n; i++) {
     // Calculate Householder vector, magnitude
-    sigma = 0;
-    v[i] = 1;
+    sigma = 0.0;
+    v[i] = 1.0;
     for (j=i+1; j<m; j++) {
       sigma += mat[i+n*j]*mat[i+n*j];
       v[j] = mat[i+n*j];
     }
-    if (sigma == 0) {
+    if (sigma<10E-14) {
       b = 0;
     } else {
       mu = sqrt(mat[i+n*i]*mat[i+n*i]+sigma);
-      if (mat[i+n*i]<=0) {
+      if (mat[i+n*i]<10E-14) {
         v[i] = mat[i+n*i]-mu;
       } else {
         v[i] = -sigma/(mat[i+n*i]+mu);
       }
       b = 2*v[i]*v[i]/(sigma+v[i]*v[i]);
-      for (j=i; j<m; j++) {
-        v[m-1-(j-i)] = v[m-1-(j-i)]/v[i];
+      for (j=i+1; j<m; j++) {
+        v[j] = v[j]/v[i];
```

Comment:
These changes would be better squashed into the previous commit.

Also, it would be easier to read/maintain with a (private) function that applies an elementary reflector, similar to [xLARF](http://www.netlib.org/lapack/explore-html/d8/d9b/group__double_o_t_h_e_rauxiliary_ga39a2b39f56ce497b218c5f48339ee712.html#ga39a2b39f56ce497b218c5f48339ee712) in LAPACK.

Also note that `xGEQRF` returns the scaling vector "`tau`", thus avoiding needing to recompute it (just extra code; we don't care about performance here). That would be a more useful interface IMO.

Let me know if you'd like to change this or just merge the branch as is? (And sorry this comment is so late.)

Reply:
I did two rounds of squashing cleanup.

Are you talking about making a common function for lines 316-328, 392-404, and 406-418?